client/app/pages/queries/QueryView.jsx

```diff
@@ -0,0 +1,112 @@
+import { Parameters } from "@/components/Parameters";
+import { TimeAgo } from "@/components/TimeAgo";
+import { currentUser } from "@/services/auth";
+import QueryPageHeader from "./components/QueryPageHeader";
+
+import { IMG_ROOT, DataSource } from "@/services/data-source";
+import QueryVisualizationTabs from "./components/QueryVisualizationTabs";
+import { EditVisualizationButton } from "@/components/EditVisualizationButton";
+import useQueryResult from "@/lib/hooks/useQueryResult";
+import { pluralize } from "@/filters";
+
+function QueryView({ query }) {
+  const canEdit = useMemo(() => currentUser.canEdit(query) || query.can_edit, [query]);
+  const parameters = useMemo(() => query.getParametersDefs(), [query]);
+  const [dataSource, setDataSource] = useState();
+  const queryResult = useMemo(() => query.getQueryResult(), [query]);
+  const queryResultData = useQueryResult(queryResult);
+
+  useEffect(() => {
+    DataSource.get({ id: query.data_source_id }).$promise.then(setDataSource);
+  }, [query]);
+  return (
+    <div className="query-page-wrapper">
+      <div className="container">
+        <QueryPageHeader query={query} />
+        <div className="query-metadata tiled bg-white p-15">
+          <EditInPlace
+            className="w-100"
+            value={query.description}
+            isEditable={canEdit}
+            editor="textarea"
+            placeholder="Add description"
+            ignoreBlanks
+          />
+          <Divider />
+          <div className="d-flex flex-wrap">
+            <div className="m-r-20 m-b-10">
+              <img src={query.user.profile_image_url} className="profile__image_thumb" alt={query.user.name} />
+              <strong>{query.user.name}</strong>
+              {" created "}
+              <TimeAgo date={query.created_at} />
+            </div>
+            <div className="m-r-20 m-b-10">
+              <img
+                src={query.last_modified_by.profile_image_url}
+                className="profile__image_thumb"
+                alt={query.last_modified_by.name}
+              />
+              <strong>{query.last_modified_by.name}</strong>
+              {" updated "}
+              <TimeAgo date={query.updated_at} />
+            </div>
+            {dataSource && (
+              <div className="m-r-20 m-b-10">
+                <img src={`${IMG_ROOT}/${dataSource.type}.png`} width="20" alt={dataSource.type} />
+                {dataSource.name}
+              </div>
+            )}
+          </div>
+        </div>
+        <div className="query-content tiled bg-white p-15 m-t-15">
+          {query.hasParameters() && <Parameters parameters={parameters} />}
+          <QueryVisualizationTabs queryResult={queryResult} visualizations={query.visualizations} />
+          <Divider />
+          <div className="d-flex align-items-center">
+            <EditVisualizationButton />
+            <Button className="icon-button hidden-xs">
```

Comment:
Could we add a `DynamicComponent` here?

Reply:
> Could we add a DynamicComponent here

Of course, no problems :+1:

Reply:
Awesome thank you!

Reply:
@jezdez It would be really nice if you show us an exact place where you need a `DynamicComponent` here and in other places (any way you wish - code snippet, suggestion or PR)

Reply:
Do you want me to send a PR to add it?

Reply:
Not necessarily. I just want to know where you need that `DynamicComponent` (and component key) - before button, after button, wrapping button (or more components), etc. - to avoid confusions.

Reply:
We erred on the side of caution and only implemented a single button like the "Edit visualization" button. I'm interested in keeping the main Redash UI consistent and not risk breaking the layout, so if restricting the type of dynamic content would help, I would encourage you to go that way.

So how about this:

Add a `<DynamicComponent name="QueryControlExtra" />` component after the "edit visualization", but before the QueryControlDropdown.

Don't wrap it so it's possible to implement a single button or a menu in case there are more custom items OR alternatively add additional `<DynamicComponent name="QueryControlDropdownExtra" />` to be able to add more menu items.